File: BinarySearch/DivideTwoInt.cpp
#include <iostream>
#include <climits>

class Solution {
public:
    int divide(int dividend, int divisor) {
        long long m = std::abs((long long) dividend);
        long long n = std::abs((long long) divisor);
        long long res = 0;

        if (m < n)
            return 0;
        while (m >= n) {
            long long t = n, p = 1;
            while (m >= (t << 1)) {
                t <<= 1;
                p <<= 1;
            }
            res += p;
            m -= t;
        }
        res = (dividend < 0) ^ (divisor < 0) ? -res : res;
        return res > INT_MAX ? INT_MAX : res;
    }
};

int main()
{
    int a = -2147483648, b = -1;

    Solution solu;
    long long res = solu.divide(a, b);
    std::cout << res << "\n";
    return 0;
}

File: BinarySearch/DivideTwoInt.py
class Solution:
    def divide(self, dividend, divisor):
        m = abs(dividend)
        n = abs(divisor)
        res = 0
        if m < n:
            return 0
        while m >= n:
            t, p = n, 1
            while (m > (t << 1)):
                t <<= 1
                p <<= 1
            res += p
            m -= t
        res = -res if (dividend < 0) ^ (divisor < 0) else res
        return res


test_cases = [[-10, 3], [10, 3], [-2147483648, -1]]

solu = Solution()
for i in range(3):
    print(solu.divide(test_cases[i][0], test_cases[i][1]))
# print(3 << 1)
